File: ime/XiaoShanIME.m
// 小删输入法 — 薄壳 IME（Objective-C）
// 职责：接收本地端口(51967)的 JSON 行指令，把文字以"预编辑(带下划线)"形式
//       实时画进当前输入框 / 定稿上屏。语音识别由 Python 引擎完成。
// 指令：{"op":"marked","text":"..."} 预编辑   {"op":"commit","text":"..."} 定稿
//       {"op":"clear"} 清空预编辑
#import <Cocoa/Cocoa.h>
#import <InputMethodKit/InputMethodKit.h>
#import <sys/socket.h>
#import <netinet/in.h>
#import <unistd.h>
#import <pthread.h>

static IMKInputController *gActive = nil;

@interface XiaoShanController : IMKInputController
@end

@implementation XiaoShanController
- (void)activateServer:(id)sender {
    [super activateServer:sender];
    gActive = self;
}
- (void)deactivateServer:(id)sender {
    if (gActive == self) gActive = nil;
    [super deactivateServer:sender];
}
// 不拦截按键；口述期间用户不打字
- (BOOL)handle:(NSEvent *)event client:(id)sender {
    return NO;
}
@end

static void applyOp(NSString *op, NSString *text) {
    dispatch_async(dispatch_get_main_queue(), ^{
        if (!gActive) return;
        id<IMKTextInput> client = [gActive client];
        if (!client) return;
        NSRange unknown = NSMakeRange(NSNotFound, NSNotFound);
        if ([op isEqualToString:@"marked"]) {
            NSRange sel = NSMakeRange(text.length, 0);
            [client setMarkedText:text selectionRange:sel replacementRange:unknown];
        } else if ([op isEqualToString:@"commit"]) {
            [client setMarkedText:@"" selectionRange:NSMakeRange(0, 0) replacementRange:unknown];
            [client insertText:text replacementRange:unknown];
        } else if ([op isEqualToString:@"clear"]) {
            [client setMarkedText:@"" selectionRange:NSMakeRange(0, 0) replacementRange:unknown];
        }
    });
}

static void serveClient(int cfd) {
    NSMutableData *buf = [NSMutableData data];
    char chunk[65536];
    NSData *nlData = [@"\n" dataUsingEncoding:NSUTF8StringEncoding];
    while (1) {
        ssize_t n = read(cfd, chunk, sizeof(chunk));
        if (n <= 0) break;
        [buf appendBytes:chunk length:(NSUInteger)n];
        NSRange nl;
        while ((nl = [buf rangeOfData:nlData options:0
                                range:NSMakeRange(0, buf.length)]).location != NSNotFound) {
            NSData *lineData = [buf subdataWithRange:NSMakeRange(0, nl.location)];
            [buf replaceBytesInRange:NSMakeRange(0, nl.location + 1) withBytes:NULL length:0];
            NSString *line = [[NSString alloc] initWithData:lineData encoding:NSUTF8StringEncoding];
            if (!line) continue;
            id obj = [NSJSONSerialization JSONObjectWithData:[line dataUsingEncoding:NSUTF8StringEncoding]
                                                    options:0 error:NULL];
            if (![obj isKindOfClass:[NSDictionary class]]) continue;
            NSString *op = obj[@"op"];
            NSString *text = obj[@"text"] ?: @"";
            if (op) applyOp(op, text);
        }
    }
    close(cfd);
}

static void *socketThread(void *arg) {
    int fd = socket(AF_INET, SOCK_STREAM, 0);
    if (fd < 0) return NULL;
    int one = 1;
    setsockopt(fd, SOL_SOCKET, SO_REUSEADDR, &one, sizeof(one));
    struct sockaddr_in addr;
    memset(&addr, 0, sizeof(addr));
    addr.sin_family = AF_INET;
    addr.sin_port = htons(51967);
    addr.sin_addr.s_addr = inet_addr("127.0.0.1");
    if (bind(fd, (struct sockaddr *)&addr, sizeof(addr)) < 0) { close(fd); return NULL; }
    if (listen(fd, 4) < 0) { close(fd); return NULL; }
    while (1) {
        int cfd = accept(fd, NULL, NULL);
        if (cfd < 0) continue;
        dispatch_async(dispatch_get_global_queue(0, 0), ^{
            serveClient(cfd);
        });
    }
    return NULL;
}

int main(int argc, char *argv[]) {
    @autoreleasepool {
        IMKServer *server = [[IMKServer alloc]
            initWithName:@"XiaoShan_IME_Connection"
            bundleIdentifier:[NSBundle mainBundle].bundleIdentifier];
        if (!server) return 1;
        pthread_t t;
        pthread_create(&t, NULL, socketThread, NULL);
        [[NSApplication sharedApplication] run];
    }
    return 0;
}

File: ime/XiaoShanIME.swift
// 小删输入法 — 薄壳 IME
// 职责：接收本地端口(51967)的 JSON 行指令，把文字以"预编辑(带下划线)"形式
//       实时画进当前输入框 / 定稿上屏。语音识别由 Python 引擎完成。
// 指令：{"op":"marked","text":"..."} 预编辑   {"op":"commit","text":"..."} 定稿
//       {"op":"clear"} 清空预编辑
import Cocoa
import InputMethodKit

let kPort: UInt16 = 51967

class XiaoShanController: IMKInputController {
    static weak var active: XiaoShanController?

    override func activateServer(_ sender: Any!) {
        super.activateServer(sender)
        XiaoShanController.active = self
    }

    override func deactivateServer(_ sender: Any!) {
        if XiaoShanController.active === self {
            XiaoShanController.active = nil
        }
        super.deactivateServer(sender)
    }

    override func handle(_ event: NSEvent!, client sender: Any!) -> Bool {
        return false   // 不拦截按键；口述期间用户不打字
    }
}

func withActiveClient(_ body: (IMKTextInput) -> Void) {
    DispatchQueue.main.async {
        if let c = XiaoShanController.active, let client = c.client() {
            body(client.takeUnretainedValue())
        }
    }
}

func applyMarked(_ text: String) {
    withActiveClient { client in
        let range = NSRange(location: (text as NSString).length, length: 0)
        client.setMarkedText(text, selectionRange: range,
                             replacementRange: NSRange(location: NSNotFound, length: NSNotFound))
    }
}

func applyCommit(_ text: String) {
    withActiveClient { client in
        client.setMarkedText("", selectionRange: NSRange(location: 0, length: 0),
                             replacementRange: NSRange(location: NSNotFound, length: NSNotFound))
        client.insertText(text, replacementRange: NSRange(location: NSNotFound, length: NSNotFound))
    }
}

func applyClear() {
    withActiveClient { client in
        client.setMarkedText("", selectionRange: NSRange(location: 0, length: 0),
                             replacementRange: NSRange(location: NSNotFound, length: NSNotFound))
    }
}

// ---------- 本地socket服务：收 JSON 行 ----------
func startSocket() {
    Thread.detachNewThread {
        let fd = socket(AF_INET, SOCK_STREAM, 0)
        guard fd >= 0 else { return }
        var one: Int32 = 1
        setsockopt(fd, SOL_SOCKET, SO_REUSEADDR, &one, socklen_t(MemoryLayout<Int32>.size))
        var addr = sockaddr_in()
        addr.sin_family = sa_family_t(AF_INET)
        addr.sin_port = kPort.bigEndian
        addr.sin_addr = in_addr(s_addr: inet_addr("127.0.0.1"))
        let bindOk = withUnsafePointer(to: &addr) {
            $0.withMemoryRebound(to: sockaddr.self, capacity: 1) {
                bind(fd, $0, socklen_t(MemoryLayout<sockaddr_in>.size))
            }
        }
        guard bindOk == 0, listen(fd, 4) == 0 else { close(fd); return }
        while true {
            let cfd = accept(fd, nil, nil)
            guard cfd >= 0 else { continue }
            Thread.detachNewThread {
                defer { close(cfd) }
                var buf = Data()
                var chunk = [UInt8](repeating: 0, count: 65536)
                while true {
                    let n = read(cfd, &chunk, chunk.count)
                    if n <= 0 { break }
                    buf.append(contentsOf: chunk[0..<n])
                    while let idx = buf.firstIndex(of: UInt8(ascii: "\n")) {
                        let lineData = buf[0..<idx]
                        buf.removeSubrange(0...idx)
                        if let line = String(data: Data(lineData), encoding: .utf8),
                           let obj = try? JSONSerialization.jsonObject(with: Data(line.utf8)) as? [String: Any],
                           let op = obj["op"] as? String {
                            let text = obj["text"] as? String ?? ""
                            switch op {
                            case "marked": applyMarked(text)
                            case "commit": applyCommit(text)
                            case "clear": applyClear()
                            default: break
                            }
                        }
                    }
                }
            }
        }
    }
}

// ---------- 启动 ----------
let connectionName = "XiaoShan_IME_Connection"
guard let server = IMKServer(name: connectionName,
                             bundleIdentifier: Bundle.main.bundleIdentifier) else {
    exit(1)
}
startSocket()
NSApplication.shared.run()
